app/oracle.py

```diff
@@ -167,50 +166,33 @@ def prompt(prompt_message, prompt_end):
 
 logging.info('Starting the main loop')
 while True:
-
-    # Get the frame and validators keys from ETH1 side
+    # Get the the epoch that is both finalized and reportable
     current_frame = oracle.functions.getCurrentFrame().call()
-    reportable_epoch = current_frame[0]
-    logging.info(f'Reportable epoch: {reportable_epoch}')
+    potentially_reportable_epoch = current_frame[0]
+    logging.info(f'Potentially reportable epoch: {potentially_reportable_epoch} (from ETH1 contract)')
+    finalized_epoch_beacon = beacon.get_finalized_epoch()
+    logging.info(f'Last finalized epoch: {finalized_epoch_beacon} (from Beacon)')
+    reportable_epoch = min(finalized_epoch_beacon, potentially_reportable_epoch) // epochs_per_frame * epochs_per_frame
```

Comment:
I'd frame it as 
reportable_epoch = `min(potentially_reportable_epoch, (finalized_epoch_beacon // epochs_per_frame) * epochs_per_frame)` which amounts to the same thing but makes sure SC is the main source of truth
After that, it needs a check that it's strictly greater than the already reported epoch, so that we won't run all the code down there when it'll cleary fail on submission.